From dede6a70766b89a0431e52bd031c9b0f15c911b8 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 31 Mar 2022 19:55:17 +0300
Subject: [PATCH] [SW-77251] common: add error apis

This patch adds common error apis required for recovery operations.
subsequent patches will make use of these functions.

Change-Id: I8da243b9dde9d7f3c35fb39857cfe3f7009b02a2
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 6ca1a62..8ebe782 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -16,6 +16,16 @@
 #include "boot_desc.h"
 #include <common.h>
 
+struct hl_rsp_err {
+	union {
+		struct {
+			uint32_t major;
+			uint32_t minor;
+		};
+		uint64_t error;
+	};
+};
+
 struct arcboot_global_s arcboot_rom_global = {
 	.config_fuse = 0,
 	.recovery_flag = 0,
@@ -43,6 +53,37 @@
 	.pci_link_up = false,
 };
 
+static struct hl_rsp_err pending_error;
+
+bool arcboot_error_inform(uint32_t major, uint32_t minor)
+{
+	/*
+	 * ensuring sending the first error occurrence
+	 * to reset the error just use the clr_on_read
+	 */
+	if (!pending_error.error) {
+		pending_error.major = major;
+		pending_error.minor = minor;
+		return true;
+	}
+
+	return false;
+}
+
+bool arcboot_error_is_pending(void)
+{
+	return !!pending_error.error;
+}
+
+uint32_t arcboot_error_read_and_clr(uint32_t *major, uint32_t *minor)
+{
+	*major = pending_error.major;
+	*minor = pending_error.minor;
+
+	pending_error.error = 0;
+	return *major;
+}
+
 #define ARCBOOT_STATUS_REG_SZ_BITS		32
 __weak void arcboot_rom_status_update(enum arcboot_rom_progress progress)
 {
diff --git a/include/arcboot.h b/include/arcboot.h
index a4d56e6..9babb18 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -214,6 +214,9 @@
 void arcboot_rom_load_recovery(void);
 int arcboot_rom_mem_init(uint64_t addr, uint32_t size);
 void arcboot_rom_verify(void);
+uint32_t arcboot_error_read_and_clr(uint32_t *major, uint32_t *minor);
+bool arcboot_error_inform(uint32_t major, uint32_t minor);
+bool arcboot_error_is_pending(void);
 
 #define CLEAR_DMA_SECURITY_PROGRESS 0x3
 
